### PR TITLE
Add additional sh lint check for incompatible sed -i cases

### PR DESCRIPTION
**Pull request by chrishollandaise**
## Changes
Fixes: #4889

Introduces additional warning lints if contributor uses only a GNU-compliant version of sed in-place command. An example of this warning as shown below (shows which scripts and line numbers).

SCR-20240120-mxxq

## Test cases
[Regexr test cases](https://regexr.com/7qntv)

## Additional Discussion
* Because linting is done on ubuntu-latest (which always used GNU grep), I made the decision to keep the command to find theses cases simple instead portable. [See this Stack Overflow thread for more information.](https://stackoverflow.com/a/6565519/9262657)
* Depending on future incompatibility problems with other coreutil commands, it is worth thinking about whether we want to have a solution that can do this sort of linting for us. Our current script linter `shellcheck` does not check for these sort of compatibility issues or even detects userspaces. [See this issue on shellcheck for more information](https://github.com/koalaman/shellcheck/issues/2902)
    * I will stipulate that this only becomes a real problem for scripts called on from `scripts/build.sh` so every time that yarn run build or any other alias runs this script. So efforts should probably be focused on just checking or minimizing the incompatibility cases.

### CODE REVIEW

File: ci/lint.sh [hunk 1]
```diff
@@ -5,5 +5,10 @@ if [ -x "$(command -v shellcheck)" ]; then
 else
 	echo "Warning: shellcheck is not installed, skipping shell scripts"
 fi
+incompatible_sed=$(! grep -RPn --color=always --exclude-dir=node_modules --exclude-dir=.git 'sed(?:\s-\w+)*\s-i(?!\S|\s"")')
```
Comment (chrishollandaise) [inline ci/lint.sh:8]: If you check the test cases, this particular regex pattern won't support all edge cases, would it be worth covering those edges? It will require more, and I probably have to consider awk to cover everything

Comment (bovine3dom) [reply ci/lint.sh:8]: I think it's probably fine as it stands; we won't be trying to trick it, it'll just be people using `sed -i` rather than `sed -i.bak`
